src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs

```diff
@@ -0,0 +1,452 @@
+                        {
+                            var ch0Index = ch0Count == 1 ? Vector.Dot(ch0Equals, _dotIndex) : byte.MaxValue;
+                            var ch1Index = ch1Count == 1 ? Vector.Dot(ch1Equals, _dotIndex) : byte.MaxValue;
+                            if (ch0Index < ch1Index)
+                            {
+                                _block = block;
+                                _index = index + ch0Index;
+                                return char0;
+                            }
+                            else
+                            {
+                                _block = block;
+                                _index = index + ch1Index;
+                                return char1;
+                            }
+                        }
+                        else
+                        {
+                            following = vectorStride;
+                        }
+                    }
+                    for (; following != 0; following--, index++)
+                    {
+                        var byteIndex = block.Array[index];
+                        if (byteIndex == byte0)
+                        {
+                            _block = block;
+                            _index = index;
+                            return char0;
+                        }
+                        else if (byteIndex == byte1)
+                        {
+                            _block = block;
+                            _index = index;
+                            return char1;
+                        }
+                    }
+                }
+            }
+        }
+
+        public int GetLength(MemoryPoolIterator2 end)
+        {
+            if (IsDefault || end.IsDefault)
+            {
+                return -1;
+            }
+
+            var block = _block;
+            var index = _index;
+            var length = 0;
+            while (true)
+            {
+                if (block == end._block)
+                {
+                    return length + end._index - index;
+                }
+                else if (block.Next == null)
+                {
+                    throw new InvalidOperationException("end did not follow iterator");
+                }
+                else
+                {
+                    length += block.End - index;
+                    block = block.Next;
+                    index = block.Start;
+                }
+            }
+        }
+
+        public string GetString(MemoryPoolIterator2 end)
+        {
+            if (IsDefault || end.IsDefault)
+            {
+                return default(string);
+            }
+            if (end._block == _block)
+            {
+                return _utf8.GetString(_block.Array, _index, end._index - _index);
+            }
+
+            var decoder = _utf8.GetDecoder();
+
+            var length = GetLength(end);
+            var charLength = length * 2;
```

Comment:
because people don't realize utf8 is actually invalid in a header, and it's not worth doing two passes to figure out what the actual char length was